### PR TITLE
[llvm][CodeGen] Update checking method of loop-carried phi in window scheduler

### PR DESCRIPTION
**Pull request by huaatian**
Added some logic to check loop-carried phis in the window scheduler. It
now includes the scenario where the preceding phi uses the virtual
register defined by the succeeding phi.

### CODE REVIEW

**Comment by huaatian** [comment]
During our internal code review, we discovered that we had missed an equivalent loop-carried phi scenario. Previously, we checked the scenario where: "The virtual register defined by the preceding phi is used by the succeeding phi." There is an equivalent scenario where: "The preceding phi uses the virtual register defined by the succeeding phi." This patch improves the check to cover this scenario.
These scenarios were included in the Hexagon test cases, but since the windowing algorithm was not executed in that test case, no issues were encountered. We have now added the corresponding test cases.
image
image

**Comment by huaatian** [comment]
Could you please review this patch? Thank you~ @arsenm

File: llvm/lib/CodeGen/WindowScheduler.cpp [hunk 1]
```diff
@@ -192,22 +192,36 @@ bool WindowScheduler::initialize() {
...
+  SmallSet<Register, 8> PrevDefs;
+  SmallSet<Register, 8> PrevUses;
+  auto IsLoopCarried = [&](MachineInstr &Phi) {
+    // Two cases are checked here: (1)The virtual register defined by the
+    // preceding phi is used by the succeeding phi;(2)The preceding phi uses the
+    // virtual register defined by the succeeding phi.
+    for (auto MO : Phi.operands()) {
+      if (!MO.isReg())
+        continue;
+      if (MO.isDef()) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/WindowScheduler.cpp:204]: You know the structure of phi so you can increment by 2 and check the single def separately

File: llvm/lib/CodeGen/WindowScheduler.cpp [hunk 2]
```diff
@@ -192,22 +192,36 @@ bool WindowScheduler::initialize() {
     return false;
   }
   // Check each MI in MBB.
-  SmallVector<Register, 8> PhiDefs;
+  SmallSet<Register, 8> PrevDefs;
+  SmallSet<Register, 8> PrevUses;
```
Comment (arsenm) [inline llvm/lib/CodeGen/WindowScheduler.cpp:196]: These sets aren't used outside of the lambda. Just make the lambda a helper function with these used locally?

File: llvm/lib/CodeGen/WindowScheduler.cpp (hunk 1 above)
Comment (huaatian) [reply llvm/lib/CodeGen/WindowScheduler.cpp]: Updated

File: llvm/lib/CodeGen/WindowScheduler.cpp (hunk 2 above)
Comment (huaatian) [reply llvm/lib/CodeGen/WindowScheduler.cpp:196]: These sets are used to record the defs and uses of phis that have already been checked in the current MBB.  
They need to be initialized and retain register information across multiple calls to the lambda. 
Therefore,  wouldn't it be better to place these sets outside the lambda?

Comment (arsenm) [reply llvm/lib/CodeGen/WindowScheduler.cpp:196]: ok

File: llvm/lib/CodeGen/WindowScheduler.cpp [hunk 3]
```diff
@@ -192,22 +192,37 @@ bool WindowScheduler::initialize() {
...
+  auto IsLoopCarried = [&](MachineInstr &Phi) {
+    // Two cases are checked here: (1)The virtual register defined by the
+    // preceding phi is used by the succeeding phi;(2)The preceding phi uses the
+    // virtual register defined by the succeeding phi.
+    auto Def = Phi.getOperand(0);
+    if (Def.isReg()) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/WindowScheduler.cpp:202]: Def must be a register. Don't try to handle invalid MIR

File: llvm/lib/CodeGen/WindowScheduler.cpp [hunk 4]
```diff
@@ -192,22 +192,37 @@ bool WindowScheduler::initialize() {
...
+      if (PrevUses.count(Def.getReg()))
+        return true;
+      PrevDefs.insert(Def.getReg());
+    }
+    for (unsigned I = 1, E = Phi.getNumOperands(); I != E; I += 2) {
+      auto Use = Phi.getOperand(I);
+      if (Use.isReg()) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/WindowScheduler.cpp:209]: Use must be a register

File: llvm/lib/CodeGen/WindowScheduler.cpp (hunk 3 above)
Comment (huaatian) [reply llvm/lib/CodeGen/WindowScheduler.cpp]: Updated

File: llvm/lib/CodeGen/WindowScheduler.cpp (hunk 4 above)
Comment (huaatian) [reply llvm/lib/CodeGen/WindowScheduler.cpp]: Updated

**Review by arsenm — APPROVED** [review]

**Pull request merged**